### PR TITLE
RFD 65 Distributed Tracing

### PR DESCRIPTION
**Pull request by rosstimothy**

### CODE REVIEW

File: rfd/0065-distributed-tracing.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 65 - Distributed Tracing
+
```
Comment (russjones) [inline rfd/0065-distributed-tracing.md:7]:
```suggestion
# Required Approvers
Engineering: @zmb3 (@fspmarshall || @espadolini)
Product: (@klizhentas || @xinding33)
```

@rosstimothy @r0mant @zmb3 Do we have any other distributed tracing experts on the team that might be good required reviewers? Maybe @codingllama?

Comment (codingllama) [reply rfd/0065-distributed-tracing.md:7]: I'm no tracing expert, I've used a couple tools over time but it's been a while. Happy to take a closer look if you like.

FWIW, I think Tim's overall proposal is a no-brainer, with OpenTelemetry and propagating spans via context.Context. Automatic tracers and boundaries are already much better than nothing; in other scenarios I think manual instrumentation is the way to get good spans. I wouldn't aim at instrumenting everything, but we can do a few codepaths we care about address others as the need arises.

**Review by zmb3 — APPROVED** [review]

File: rfd/0065-distributed-tracing.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+* Generate and export spans for Teleport service boundaries and core components
+  (e.g. http client/server, grpc client/server, ssh client/server, cache, backend)
+* Provide a solid foundation for how to use tracing within Teleport so that more areas of the
+  code can easily be instrumented in the future
+
+### Non-Goals
+
+* Adding tracing to the entire codebase all at once
+* Replace existing logging, metrics
```
Comment (jakule) [inline rfd/0065-distributed-tracing.md:49]: I do understand that replacing existing metrics is not a goal of this RFD, but from what I remember Teleport is currently using Prometheus for some metrics. What are the pros/cons of adding Open Telemetry vs extending our already existing Prometheus metrics?

File: rfd/0065-distributed-tracing.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+added to other areas of the codebase, using the code already instrumented as a guide for best practices.
+
+All spans created to instrument Teleport shall follow the guidelines laid out in the
+[spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span).
+Likewise, all attributes shall follow the [guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attribute)
+to provide consistency within Teleport and across other services.
+
+#### Exporting Spans
+
+All spans created by Teleport will be exported in the
```
Comment (jakule) [inline rfd/0065-distributed-tracing.md:164]: From what I see Open Telemetry implements client-server model where the client (Teleport in our case) is responsible for exporting the metrics. From what I know this approach has one major problem. If the system is under heavy load then metrics tracking can significantly slow the system. From what I remember this is one of the main reason why Prometheus for example decided to just expose an endpoint and let the server to scrape the metrics whatever needed. 

Do you know how this problem is solved (or maybe still exist) in Open Telemetry? Do we have any metrics that would show how Teleport would behave in such a scenario?

File: rfd/0065-distributed-tracing.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 65 - Distributed Tracing
+
+## What
+
+Add distributed tracing to Teleport by leveraging [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go)
```
[the remark this answers is not in the document]
Comment (rosstimothy) [reply rfd/0065-distributed-tracing.md:10]: I did look into other solutions like OpenTracing, OpenCensus, runtime/trace and x/net/trace. OpenTelemetry seems to be the standard everyone is adopting. The OTLP format was also a big reason for selecting using OpenTelemetry. By exporting spans in OTLP format, we don't have to write any code that is specific to a particular telemetry backend. We can simply export OTLP spans and all users should be able to import it either directly into their telemetry backend or make use of the OpenTelemetry collector to convert it for their telemetry backend.

File: rfd/0065-distributed-tracing.md (hunk 2 above)
Comment (rosstimothy) [reply rfd/0065-distributed-tracing.md:49]: This RFD is specifically for adding and exporting distributed tracing data, which is something that Prometheus is not capable of doing. The mention of metrics here is to indicate that this RFD does not intend to address the possibility of replacing Prometheus with OpenTelemetry metrics.

File: rfd/0065-distributed-tracing.md (hunk 3 above)
Comment (rosstimothy) [reply rfd/0065-distributed-tracing.md:164]: You are correct spans would be collected and exported from Teleport to an OpenTelemetry Collector. While I haven't done any testing under heavy load yet, this situation could potentially still cause exporting spans to be delayed. However, as stated in this RFD, the number of spans being sampled and exported will be determined by `sampling_rate_per_hundred`. Setting a small value is something that should be encouraged in almost all scenarios. 

I think that the sampling rate in conjunction with asynchronously exporting spans with a [BatchSpanProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#NewBatchSpanProcessor) should allow us to be able to continue to export spans even under load.

**Review by jakule — APPROVED** [review]

File: rfd/0065-distributed-tracing.md (hunk 2 above)
Comment (strideynet) [inline rfd/0065-distributed-tracing.md:49]: As a side note on this, I largely agree with avoiding changing anything re: metrics. My recent experiences with Otel Tracing have been extremely positive, but the Metrics API is still incredibly immature and making a lot of breaking changes, and I've also had a few issues/bugs with their Metrics SDK. We can probably safely revisit the metrics side and Otel in six to twelve months time.

File: rfd/0065-distributed-tracing.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+and [`SpanExporter`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#SpanExporter) which would only record 10% of
+all spans and export them to the specified url.
+
+
+#### Exporting Spans from tsh/tctl/tbot
+
+In order to propagate spans that originated from `tctl`, `tsh`, or `tbot` we have two options:
+
+1) Export the spans directly into the telemetry backend
+2) Make the Auth Server a span forwarder
```
Comment (strideynet) [inline rfd/0065-distributed-tracing.md:195]: I strongly agree with pursuing the second option here. Exporting spans directly to the telemetry backend raises a lot of issues around preventing abuse and authentication, especially if we want to use this within Teleport cloud.

File: rfd/0065-distributed-tracing.md (hunk 1 above)
Comment (strideynet) [reply rfd/0065-distributed-tracing.md:7]: I'd be happy to come onto this as an approver/reviewer, I've done a fair bit with Observability in previous roles.

**Review by strideynet — APPROVED** [review]
This is looking really good. I've found tracing to be an incredibly useful tool in working through performance issues at scale, and it starts moving us towards the Otel ecosystem which is something that I believe will only grow in popularity as it matures over the next year or two.

Whilst there will be plenty of refactoring needed to implement this (e.g properly propagating context across Teleport's codebase), most of this work is really long overdue anyway.

**Review by fspmarshall — APPROVED** [review]
LGTM.  I don't know how feasible it would be, but I think it would be very nice to have some mechanism to export a trace locally without setting up a cluster-level export destination.  I don't know if there is a commonly agreed upon file format for traces that could be exported to, but even just being able to run an export destination locally in docker when running something like `tsh ls --trace` would be nice.  Would be especially useful if we're trying to assist people in debugging an issue when they don't currently have a cluster-level tracing endpoint enabled.

**Review by espadolini — APPROVED** [review]
A couple things that were discussed with @gozer and @rosstimothy in light of gravitational/cloud#1765

File: rfd/0065-distributed-tracing.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+different backends. Due to the fact that different Teleport customers will have different telemetry backends this
+provides us a way to export traces in such a way that they can be easily consumed all customers. Most of the major
+telemetry backends have already, or are working on adding support to consume OTLP spans. Customers also have the option
+to use the [OpenTelemetry Collector (otel-collector)](https://github.com/open-telemetry/opentelemetry-collector) to
+receive, process, and export the traces from Teleport to their desired backend. This is particularly useful for telemetry
+backends which don't currently support OTLP. The otel-collector could be configured to retrieve spans from Teleport and
+then forward them the backend in a format that the backend supports.
+
+
+##### Tracing Configuration
```
Comment (xinding33) [inline rfd/0065-distributed-tracing.md:180]: Where do we plan to add these configuration options? `teleport.yaml`?

**Review by xinding33 — APPROVED** [review]
I'm a fan of using `OpenTelemetry` as it allows for easy export of spans for analysis. ~~Do we want to have someone from the Cloud team take a look given that we'd need to instrument this on Teleport Cloud?~~

Edit: Cloud has already reviewed this and staging is already ingesting spans.

**Comment by gozer** [comment]
> It doesn't look like @gozer has approved this RFD, 

Didn't exactly feel appropriate for me to do so, but I certainly could look again and do so.

> but I know he has looked it over in the past. 

I sure have

> We have been working together to get Cloud to ingest spans, which is currently already happening in staging.

Will be also running in production this week, in theory.

> In fact, it was his suggestion to tag the forwarded spans so Cloud could potentially omit some spans from their ingestion pipeline.

Yes, I remember discussing two distinct possiblilties and factored them in the cloud's design. One was to allow differentiation of client initiated traces (`tsh --trace`) and the possiblity of customer forwarded traces as well.

As an aside, if the plan is to support sending trace data to arbitrary 3rd party destinations, especially by tenant, considerations would have to be made, the current OpenTelemetry [Cloud-RDF#0025](https://github.com/gravitational/cloud/blob/master/rfd/0025-Distributed-Tracing-Ingestion-Service.md#non-goals) had that type of use-case as an explicit Non-Goal.

Could certainly be done, but not right away with what we currently have in place.

**Pull request merged**